// daemo-engine/public/utils/type-utils.js
"use strict";
/**
 * Utility functions for TypeScript type extraction and conversion
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTypeString = getTypeString;
exports.getParameterTypes = getParameterTypes;
exports.getReturnType = getReturnType;
exports.getTypeName = getTypeName;
exports.isAsyncFunction = isAsyncFunction;
exports.getActualReturnType = getActualReturnType;
exports.toPascalCase = toPascalCase;
exports.toCamelCase = toCamelCase;
exports.isValidTypeScriptIdentifier = isValidTypeScriptIdentifier;
/**
 * Convert a JavaScript/TypeScript type to a string representation
 */
function getTypeString(value) {
    if (value === null)
        return "null";
    if (value === undefined)
        return "undefined";
    const type = typeof value;
    if (type === "object") {
        if (Array.isArray(value)) {
            if (value.length > 0) {
                const elementType = getTypeString(value[0]);
                return `${elementType}[]`;
            }
            return "any[]";
        }
        if (value instanceof Date)
            return "Date";
        return "object";
    }
    return type;
}
/**
 * Extract parameter types from a function using TypeScript metadata
 */
function getParameterTypes(target, propertyKey) {
    const types = Reflect.getMetadata("design:paramtypes", target, propertyKey) || [];
    return types;
}
/**
 * Extract return type from a function using TypeScript metadata
 */
function getReturnType(target, propertyKey) {
    return Reflect.getMetadata("design:returntype", target, propertyKey);
}
/**
 * Get TypeScript type name from constructor
 */
function getTypeName(type) {
    if (!type || type === Object)
        return "any";
    // Handle primitive types
    if (type === String)
        return "string";
    if (type === Number)
        return "number";
    if (type === Boolean)
        return "boolean";
    if (type === Date)
        return "Date";
    if (type === Array)
        return "any[]";
    // Handle Promise
    if (type.name === "Promise")
        return "Promise<any>";
    // Return the class name
    return type.name || "any";
}
/**
 * Check if a function is async
 */
function isAsyncFunction(fn) {
    return fn.constructor.name === "AsyncFunction";
}
/**
 * Extract the actual return type from async functions
 */
function getActualReturnType(returnType) {
    if (!returnType)
        return "any";
    const typeName = getTypeName(returnType);
    if (typeName.startsWith("Promise<")) {
        return typeName.substring(8, typeName.length - 1);
    }
    return typeName;
}
/**
 * Convert camelCase to PascalCase
 */
function toPascalCase(str) {
    return str.charAt(0).toUpperCase() + str.slice(1);
}
/**
 * Convert PascalCase to camelCase
 */
function toCamelCase(str) {
    return str.charAt(0).toLowerCase() + str.slice(1);
}
/**
 * Check if a string is a valid TypeScript identifier
 */
function isValidTypeScriptIdentifier(name) {
    return /^[a-zA-Z_$][a-zA-Z0-9_$]*$/.test(name);
}
