// daemo-engine/public/handler/function-handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DaemoFunctionHandler = void 0;
exports.getGlobalFunctionHandler = getGlobalFunctionHandler;
class DaemoFunctionHandler {
    constructor() {
        this.functions = new Map();
    }
    registerDaemoFunction(func) {
        this.functions.set(func.name, func);
    }
    async executeFunction(functionName, inputJson) {
        console.log(`[Daemo SDK] Executing '${functionName}' with input: ${inputJson}`);
        try {
            const func = this.functions.get(functionName);
            if (!func) {
                throw new Error(`Function '${functionName}' not found.`);
            }
            const parsedArgs = JSON.parse(inputJson);
            const validationResult = func.inputSchema.safeParse(parsedArgs);
            if (!validationResult.success) {
                throw new Error(`Invalid input: ${validationResult.error.message}`);
            }
            const result = await func.handler(validationResult.data);
            func.outputSchema.parse(result); // Validate output as well
            const resultJson = result === undefined ? "null" : JSON.stringify(result);
            return { success: true, resultJson, errorMessage: "" };
        }
        catch (error) {
            console.error(`[Daemo SDK] Error executing function '${functionName}':`, error);
            return {
                success: false,
                resultJson: "",
                errorMessage: error.message || String(error),
            };
        }
    }
}
exports.DaemoFunctionHandler = DaemoFunctionHandler;
// Global handler instance
let globalFunctionHandler = null;
function getGlobalFunctionHandler() {
    if (!globalFunctionHandler) {
        globalFunctionHandler = new DaemoFunctionHandler();
    }
    return globalFunctionHandler;
}
