// daemo-engine/public/index.js
"use strict";
/**
 * Daemo SDK for Node.js
 *
 * Main entry point for the SDK
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getGlobalFunctionHandler = exports.DaemoClient = exports.DaemoHostedConnection = exports.FromDaemoContext = exports.ParameterDescription = exports.DaemoSchema = exports.DaemoFunction = exports.DaemoBuilder = void 0;
// Builder
var daemo_builder_1 = require("./builder/daemo-builder");
Object.defineProperty(exports, "DaemoBuilder", { enumerable: true, get: function () { return daemo_builder_1.DaemoBuilder; } });
// Decorators
var daemo_function_1 = require("./decorators/daemo-function");
Object.defineProperty(exports, "DaemoFunction", { enumerable: true, get: function () { return daemo_function_1.DaemoFunction; } });
Object.defineProperty(exports, "DaemoSchema", { enumerable: true, get: function () { return daemo_function_1.DaemoSchema; } });
Object.defineProperty(exports, "ParameterDescription", { enumerable: true, get: function () { return daemo_function_1.ParameterDescription; } });
Object.defineProperty(exports, "FromDaemoContext", { enumerable: true, get: function () { return daemo_function_1.FromDaemoContext; } });
// Connection
var daemo_hosted_connection_1 = require("./connection/daemo-hosted-connection");
Object.defineProperty(exports, "DaemoHostedConnection", { enumerable: true, get: function () { return daemo_hosted_connection_1.DaemoHostedConnection; } });
var daemo_client_1 = require("./connection/daemo-client");
Object.defineProperty(exports, "DaemoClient", { enumerable: true, get: function () { return daemo_client_1.DaemoClient; } });
// Types
__exportStar(require("./types/metadata"), exports);
__exportStar(require("./types/config"), exports);
__exportStar(require("./types/results"), exports);
// Handler
var function_handler_1 = require("./handler/function-handler");
Object.defineProperty(exports, "getGlobalFunctionHandler", { enumerable: true, get: function () { return function_handler_1.getGlobalFunctionHandler; } });
