// daemo-engine/public/connection/grpc-manager.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.grpcManager = void 0;
const proto_loader_1 = require("../utils/proto-loader");
const grpc_js_1 = require("@grpc/grpc-js");
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
class GrpcManager {
    constructor() {
        this.gatewayClient = null;
        this.agentClient = null;
        this.gatewayUrl = "localhost:50052";
        this.agentUrl = "localhost:50052";
    }
    static getInstance() {
        if (!GrpcManager.instance) {
            GrpcManager.instance = new GrpcManager();
        }
        return GrpcManager.instance;
    }
    configure(gatewayUrl, agentUrl) {
        this.gatewayUrl = gatewayUrl;
        this.agentUrl = agentUrl;
        this.gatewayClient = null;
        this.agentClient = null;
    }
    parseGrpcTarget(url) {
        try {
            const parsed = new URL(url);
            return `${parsed.hostname}:${parsed.port}`;
        }
        catch (error) {
            return url.replace(/^https?:\/\//, "");
        }
    }
    createCredentials(url) {
        const isSecure = url.startsWith("https://");
        if (isSecure) {
            console.log("[Daemo SDK] Creating secure SSL credentials.");
            const caCertB64 = process.env.TLS_CA_CERT_B64;
            if (caCertB64) {
                console.log("[Daemo SDK] Creating SSL credentials using Root CA from environment variable.");
                const rootCert = Buffer.from(caCertB64, "base64");
                return grpc_js_1.credentials.createSsl(rootCert);
            }
            try {
                const rootCertPath = path.resolve(process.cwd(), "ca.pem");
                if (fs.existsSync(rootCertPath)) {
                    console.log(`[Daemo SDK] Creating SSL credentials using ca.pem file from ${rootCertPath}`);
                    const rootCert = fs.readFileSync(rootCertPath);
                    return grpc_js_1.credentials.createSsl(rootCert);
                }
            }
            catch (e) {
                // Ignore if file doesn't exist and fall through
            }
            console.warn("[Daemo SDK] No custom CA found. Falling back to default system SSL credentials. This may fail if the server uses a self-signed or private CA.");
            return grpc_js_1.credentials.createSsl();
        }
        return grpc_js_1.credentials.createInsecure();
    }
    getGatewayClient() {
        if (!this.gatewayClient) {
            const target = this.parseGrpcTarget(this.gatewayUrl);
            const creds = this.createCredentials(this.gatewayUrl);
            this.gatewayClient = new proto_loader_1.daemoProto.DaemoGateway(target, creds, {
                "grpc.keepalive_time_ms": 30000,
                "grpc.keepalive_timeout_ms": 10000,
                "grpc.keepalive_permit_without_calls": 1,
            });
        }
        return this.gatewayClient;
    }
    getAgentClient() {
        if (!this.agentClient) {
            const target = this.parseGrpcTarget(this.agentUrl);
            const creds = this.createCredentials(this.agentUrl);
            this.agentClient = new proto_loader_1.daemoProto.DaemoAgent(target, creds);
        }
        return this.agentClient;
    }
    resetClients() {
        if (this.gatewayClient) {
            this.gatewayClient.close();
            this.gatewayClient = null;
        }
        if (this.agentClient) {
            this.agentClient.close();
            this.agentClient = null;
        }
        console.log("[Daemo SDK] gRPC clients have been reset.");
    }
}
exports.grpcManager = GrpcManager.getInstance();
